File: cole/day18_last-stone-weight/last-stone-weight.py
from heapq import heapify, heappop, heappush


class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:
        stones = [-i for i in stones]
        heapify(stones)
        while len(stones) > 1:
            a = heappop(stones)
            b = heappop(stones)
            if a != b:
                heappush(stones, -abs(a - b))
        return -stones[0] if stones else 0

File: gabe/last_stone_weight.cpp
class Solution {
public:
    priority_queue<int> pq;

    int lastStoneWeight(vector<int>& stones) {
        for(int stone: stones) {
            pq.push(stone);
        }

        int x, y;
        while (pq.size() > 1) {
            y = pq.top();
            pq.pop();
            x = pq.top();
            pq.pop();

            if (y > x) pq.push(y - x);
        }

        if (pq.size() == 1) return pq.top();

        return 0;
    }


};